### ISSUE TITLE
Allow selecting a protobuf .proto or .desc schema when parsing protobuf data

### ISSUE THREAD

**Issue by pimterry** [httptoolkit/httptoolkit#559]
We need to:

* Provide a UI to select these files, if one isn't specified (should this live in the Pro settings, mapping URLs to schemas? Or be selected from a header in the body section directly?)
* Allow selecting and/or auto-detecting (from content-type headers) the message type, as an inline option on the message content (or just test against all message types, if there aren't too many?)
* Auto-detect desc URLs if included in the content-type and offer to fetch them

Supporting both proto & desc files isn't necessarily required, but it might play nicer with autogeneration of .proto files for reverse engineering work if we can handle both, and it's just generally more convenient.

---

_Does this affect you too? Click below and add a :+1: to vote for this and help decide where HTTP Toolkit goes next, or [go vote on the other most popular ideas so far](https://github.com/httptoolkit/feedback/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc)._

**Comment by griffinli** [comment]
What's the priority of this?

**Comment by pimterry** [comment]
I'd like to add this, but there's a lot of things I'd love to do and it's quite a long way down the list. PRs welcome! Everything involved is open source and this is a purely UI change in https://github.com/httptoolkit/httptoolkit-ui/. Note that HTTP Toolkit Pro is free for all contributors to all https://github.com/httptoolkit repos :smile:

**Comment by griffinli** [comment]
Is there already code inside the app to parse a valid .proto file?

**Comment by pimterry** [comment]
@griffinli No, I don't think so. You can find most of the existing protobuf decoding logic [here](https://github.com/httptoolkit/httptoolkit-ui/blob/main/src/util/protobuf.ts). You might also be interested in the editor language config [here](https://github.com/httptoolkit/httptoolkit-ui/blob/9f03ec16576726bcda3c00faeb50592976f516fe/src/components/editor/monaco.ts#L53-L77) and the entrypoint to the body formatting [here](https://github.com/httptoolkit/httptoolkit-ui/blob/9f03ec16576726bcda3c00faeb50592976f516fe/src/services/ui-worker-formatters.ts#L108-L111).
